### PR TITLE
HIVE-27671 - Implement array_append UDF to append an element to array

### PR DESCRIPTION
**Pull request by tarak271**
### What changes were proposed in this pull request?

Implement array_append function in Hive

### Why are the changes needed?

This enhancement is already implemented in Spark

### Does this PR introduce _any_ user-facing change?

No
### Is the change a dependency upgrade?

No

### How was this patch tested?

Created Junit tests as well as qtests as part of this change

### CODE REVIEW

File: ql/src/test/org/apache/hadoop/hive/ql/udf/generic/TestGenericUDFArrayAppend.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    m3.put(new Text("a"), new IntWritable(1));
+
+    runAndVerify(asList(m1, m3, m2, m3, m1), m2, asList(m1, m3, m2, m3, m1, m2));
+  }
+
+  private void runAndVerify(List<Object> actual, Object element, List<Object> expected)
+      throws HiveException {
+    GenericUDF.DeferredJavaObject[] args = { new GenericUDF.DeferredJavaObject(actual), new GenericUDF.DeferredJavaObject(element) };
+    List<Object> result = (List<Object>) udf.evaluate(args);
+    if(expected == null){
```
Comment (SourabhBadhya) [inline ql/src/test/org/apache/hadoop/hive/ql/udf/generic/TestGenericUDFArrayAppend.java:145]: nit: Spacing

File: ql/src/test/org/apache/hadoop/hive/ql/udf/generic/TestGenericUDFArrayAppend.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+      Assert.assertEquals(result,null);
+    }
+    else {
```
Comment (SourabhBadhya) [inline ql/src/test/org/apache/hadoop/hive/ql/udf/generic/TestGenericUDFArrayAppend.java:148]: nit:` } else {`

File: ql/src/test/queries/clientpositive/udf_array_append.q [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+SELECT array_append(array(1, 2, 3), CAST(null AS int));
+
+SELECT array_append(array(1.1234567890, 2.234567890, 3.34567890, null, 3.3456789, 2.234567,1.1234567890),1.1234567890);
+
+SELECT array_append(array(11234567890, 2234567890, 334567890, null, 11234567890, 2234567890, 334567890, null),11234567890);
+
+SELECT array_append(array(array("a","b","c","d"),array("a","b","c","d"),array("a","b","c","d","e"),null,array("e","a","b","c","d")),array("a","b","c","d"));
+
+SELECT array_append(array("aa","bb","cc"),"cc");
```
Comment (SourabhBadhya) [inline ql/src/test/queries/clientpositive/udf_array_append.q:27]: Can we have a failure case, when the first argument is not an array type?

File: ql/src/test/org/apache/hadoop/hive/ql/udf/generic/TestGenericUDFArrayAppend.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    Object i3 = asList(new Text("a"), new DoubleWritable(3.1415), new DateWritableV2(Date.of(2015, 5, 25)),
+        asList(new IntWritable(1), new IntWritable(3), new IntWritable(2), new IntWritable(5)));
+
+    Object i4 = asList(new Text("a"), new DoubleWritable(3.1415), new DateWritableV2(Date.of(2015, 5, 25)),
+        asList(new IntWritable(1), new IntWritable(3), new IntWritable(2), new IntWritable(4)));
+
+    runAndVerify(asList(i1, i3, i2, i3, i4), i2, asList(i1, i3, i2, i3, i4, i2));
+  }
+
+  @Test public void testxMap() throws HiveException {
```
Comment (SourabhBadhya) [inline ql/src/test/org/apache/hadoop/hive/ql/udf/generic/TestGenericUDFArrayAppend.java:111]: Nit: Typo, should be `testMap`

File: ql/src/test/org/apache/hadoop/hive/ql/udf/generic/TestGenericUDFArrayAppend.java (hunk 1 above)
Comment (tarak271) [reply ql/src/test/org/apache/hadoop/hive/ql/udf/generic/TestGenericUDFArrayAppend.java]: Done

File: ql/src/test/queries/clientpositive/udf_array_append.q (hunk 3 above)
Comment (tarak271) [reply ql/src/test/queries/clientpositive/udf_array_append.q:27]: Added two negative test cases

File: ql/src/test/org/apache/hadoop/hive/ql/udf/generic/TestGenericUDFArrayAppend.java (hunk 4 above)
Comment (tarak271) [reply ql/src/test/org/apache/hadoop/hive/ql/udf/generic/TestGenericUDFArrayAppend.java]: Done

**Review by SourabhBadhya — APPROVED** [review]
LGTM +1

**Pull request merged**